### ISSUE TITLE
[BUG] - MLFlow 2.7.0 produces unpredictable behaviour when patching methods

### ISSUE THREAD

**Issue by Carl132** [mlflow/mlflow#9623]
### Issues Policy acknowledgement

- [X] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Willingness to contribute

No. I cannot contribute a bug fix at this time.

### MLflow version

- Client: 2.7.0
- azureml-mlflow: 1.53.0
- mlflow-skinny: 2.7.0

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 20.04.6
- **Python version**: 3.11.4
- Running in an Azure DevOps Compute Agent

### Describe the problem

Specifically when using `mlflow==2.7.0`, an issue occurs when running pytest with tests that patch the `mlflow.sklearn.X` method. Specifically for this bug report, it's raised when invoking the `mlflow.sklearn.load_model` method.

When attempting to run the following test:
```
import sys
from unittest.mock import patch
import pytest
import mlflow

# Test that fails
@patch("src.batch_inference.scripts.batch_data_prep.mlflow.sklearn.load_model")
def test_batch_data_prep_load_invalid_transformer(load_patch):
    """Test handling of loading in a transformer that doesn't exist."""
    load_patch.side_effect = mlflow.exceptions.RestException(
        json={"error_code": "RESOURCE_DOES_NOT_EXIST"}
    )
    with pytest.raises(ValueError):
        main()
```
with the context of creating unit tests for an AzureML Pipeline Component code, 
when trying to test the following method:

```
def main():
    """Load in any transformers as well as data, and preprocess it."""
    try:
        print("Loading in transformer from training")
        loaded_transformer = mlflow.sklearn.load_model(
            model_uri=f"models:/some_incorrect_model_name/latest"
        )
    except RestException as error_out:
        if "RESOURCE_DOES_NOT_EXIST" in error_out.error_code:
            print("No transformer named some_incorrect_model_name exists")
            raise FileNotFoundError("Failed to load some_incorrect_model_name")
        else:
            print("Unknown error")
            raise ValueError(f"Failed to load transformer: {error_out}")
```
I'm met with the following error:

```
load_patch = <MagicMock name='load_model' id='139966890346960'>

tox_test_proj/tests/batch_inference/scripts/test_batch_data_prep.py:150: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

tox_test_proj/src/batch_inference/scripts/batch_data_prep.py:84: in main
    loaded_transformer = mlflow.sklearn.load_model(
.tox/py311/lib/python3.11/site-packages/mlflow/sklearn/__init__.py:610: in load_model
    local_model_path = _download_artifact_from_uri(artifact_uri=model_uri, output_path=dst_path)
.tox/py311/lib/python3.11/site-packages/mlflow/store/artifact/utils/models.py:32: in _get_latest_model_version
    latest = client.get_latest_versions(name, None if stage is None else [stage])
.tox/py311/lib/python3.11/site-packages/mlflow/tracking/client.py:2429: in get_latest_versions
    return self._get_registry_client().get_latest_versions(name, stages)
.tox/py311/lib/python3.11/site-packages/mlflow/tracking/_model_registry/client.py:139: in get_latest_versions
    return self.store.get_latest_versions(name, stages)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <mlflow.store.model_registry.file_store.FileStore object at 0x7f4c94c691d0>
name = 'incorrect_name', stages = None

    def get_latest_versions(self, name, stages=None):
        """
        Latest version models for each requested stage. If no ``stages`` argument is provided,
        returns the latest version for each stage.
    
        :param name: Registered model name.
        :param stages: List of desired stages. If input list is None, return latest versions for
                       each stage.
        :return: List of :py:class:`mlflow.entities.model_registry.ModelVersion` objects.
        """
        registered_model_path = self._get_registered_model_path(name)
        if not exists(registered_model_path):
>           raise MlflowException(
                f"Registered Model with name={name} not found",
                RESOURCE_DOES_NOT_EXIST,
            )
E           mlflow.exceptions.MlflowException: Registered Model with name=incorrect_name not found
```
As we can see at the top of the stdout for the method's error, `self = <mlflow.store....` which shows that the patch did not apply properly. What's more strange is that this behaviour only occurs for the first instance of patching a method; all future patches apply successfully and the tests pass.

This behaviour is reproducable in that I can augment the order of my unit tests and it's always the first patch that fails, and all future patches will succeed.

TL;DR:
- Desired behaviour: patch applies to `mlflow.sklearn.load_model` and `RestException({"error_code": "RESOURCE_DOES_NOT_EXIST"})` is raised
- Current behaviour: `mlflow.sklearn.load_model` is accessed and the patch is not invoked. This **only** occurs for the first unit test that attempts to make a patch in the test suite.
- Downgrading to `mlflow==2.6.0` totally fixes this problem.

Note: I haven't tested this issue for any methods outside of `mlflow.sklearn.load_model` and `mlflow.sklearn.log_model`, but it's an issue for both of these.

### Tracking information

```shell
REPLACE_ME
```

### Code to reproduce issue

```test.py
import pytest
from unittest.mock import patch
import mlflow

@patch("src.batch_inference.scripts.batch_data_prep.mlflow.sklearn.load_model")
def test_batch_data_prep_load_invalid_transformer(load_patch):
    """Test handling of loading in a transformer that doesn't exist."""
    load_patch.side_effect = mlflow.exceptions.RestException(
        json={"error_code": "RESOURCE_DOES_NOT_EXIST"}
    )
    with pytest.raises(ValueError):
        main()
```

```my_file.py
import mlflow

def main(input_transformer_name="some_incorrect_model"):
    """Load in any transformers as well as data, and preprocess it."""
    try:
        print("Loading in transformer from training")
        loaded_transformer = mlflow.sklearn.load_model(
            model_uri=f"models:/{input_transformer_name}/latest"
        )
    except RestException as error_out:
        if "RESOURCE_DOES_NOT_EXIST" in error_out.error_code:
            print(f"No transformer named {input_transformer_name} exists")
            raise FileNotFoundError(f"Failed to load {input_transformer_name}")
        else:
            print("Unknown error")
            raise ValueError(f"Failed to load transformer: {error_out}")
```

### Stack trace

```
=================================== FAILURES ===================================
________________ test_batch_data_prep_load_invalid_transformer _________________

load_patch = <MagicMock name='load_model' id='139966890346960'>

    @patch("src.batch_inference.scripts.batch_data_prep.mlflow.sklearn.load_model")
    def test_batch_data_prep_load_invalid_transformer(load_patch):
        """Test handling of loading in a transformer that doesn't exist."""
        load_patch.side_effect = mlflow.exceptions.RestException(
            json={"error_code": "RESOURCE_DOES_NOT_EXIST"}
        )
        with pytest.raises(ValueError):
>           main()

tox_test_proj/tests/batch_inference/scripts/test_batch_data_prep.py:150: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tox_test_proj/src/batch_inference/scripts/batch_data_prep.py:84: in main
    loaded_transformer = mlflow.sklearn.load_model(
.tox/py311/lib/python3.11/site-packages/mlflow/sklearn/__init__.py:610: in load_model
    local_model_path = _download_artifact_from_uri(artifact_uri=model_uri, output_path=dst_path)
.tox/py311/lib/python3.11/site-packages/mlflow/tracking/artifact_utils.py:100: in _download_artifact_from_uri
    return get_artifact_repository(artifact_uri=root_uri).download_artifacts(
.tox/py311/lib/python3.11/site-packages/mlflow/store/artifact/artifact_repository_registry.py:115: in get_artifact_repository
    return _artifact_repository_registry.get_artifact_repository(artifact_uri)
.tox/py311/lib/python3.11/site-packages/mlflow/store/artifact/artifact_repository_registry.py:72: in get_artifact_repository
    return repository(artifact_uri)
.tox/py311/lib/python3.11/site-packages/mlflow/store/artifact/models_artifact_repo.py:44: in __init__
    uri = ModelsArtifactRepository.get_underlying_uri(artifact_uri)
.tox/py311/lib/python3.11/site-packages/mlflow/store/artifact/models_artifact_repo.py:78: in get_underlying_uri
    (name, version) = get_model_name_and_version(client, uri)
.tox/py311/lib/python3.11/site-packages/mlflow/store/artifact/utils/models.py:94: in get_model_name_and_version
    return model_name, str(_get_latest_model_version(client, model_name, model_stage))
            )
E           mlflow.exceptions.MlflowException: Registered Model with name=incorrect_name not found

.tox/py311/lib/python3.11/site-packages/mlflow/store/model_registry/file_store.py:362: MlflowException
```

### Other info / logs

```
REPLACE_ME
```

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [X] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by BenWilson2** [comment]
Since you're not directly patching the function within mlflow, but rather patching code from within your repository that contains this, this PR may have affected your workflow: https://github.com/mlflow/mlflow/commit/8949d84b0ef87287c57def8272cf7d9870945a99 
You may need to directly patch the method due to the LazyLoader() implementation.

**Comment by serena-ruan** [comment]
This also affects automl tests, and this sounds like a bug according to the comment in https://github.com/chenmoneygithub/mlflow/blob/8f46545582126cd3e33ab07e8aa39e191970b301/tests/tracking/fluent/test_fluent_autolog.py#L107-L108 @chenmoneygithub Could you take a look?

**Comment by chenmoneygithub** [comment]
Yea, lazy loading is not working well with MagicMock for some mystery reason. 

@Carl132 The simplest fix is loading the `mlflow.sklearn` module explicitly before the test, similar to [this](https://github.com/chenmoneygithub/mlflow/blob/8f46545582126cd3e33ab07e8aa39e191970b301/tests/tracking/fluent/test_fluent_autolog.py#L107-L108). Could you check if that works for you?

**Comment by serena-ruan** [comment]
@chenmoneygithub Do you know why it causes problem with patching? I think this is a hard-to-debug issue if people don't know this, if we don't know how to fix it could we revert the change? cc @harupy

**Comment by Carl132** [comment]
Sorry for the accidental close

This issue is very much still persistent with v2.8.1. I have used a version of what @chenmoneygithub suggested, but as I was patching as decorators to my function, I've added a test before my initial failing one as follows:

```
@patch("<path_to_module>.mlflow.sklearn.save_model")
def test_dummy_mlflow_call(mock_save):
    """A dummy mlflow call to circumvent a bug - need to load in the module."""
    mlflow.sklearn
```

And so this will successfully load in the `mlflow.sklearn` module, trying to patch the `save_model` method (but doesn't due to lazyloading) and then my subsequent tests can run properly. 

Leaving this here in case others come into this issue, but again this is a workaround and the issue isn't fixed.
